sympy/matrices/expressions/applyfunc.py

```diff
@@ -65,3 +74,26 @@ def doit(self, **kwargs):
             return expr.applyfunc(self.function)
         else:
             return self
+
+    def _entry(self, i, j, **kwargs):
+        return self.function(self.expr[i, j])
+
+    def _eval_derivative_matrix_lines(self, x):
+        d = Dummy("d")
+        function = self.function(d)
+        fdiff = function.fdiff()
+        if isinstance(fdiff, Function):
+            fdiff = type(fdiff)
+        else:
+            fdiff = Lambda(d, function.diff(d))
+        lr = self.expr._eval_derivative_matrix_lines(x)
+        if 1 in self.shape:
+            # Vector:
+            ewdiff = ElementwiseApplyFunction(fdiff, self.expr)
+            ewdiff = diagonalize_vector(ewdiff)
+            # is it a vector or a matrix or a scalar?
+            lr[0].first *= ewdiff
+            return lr
+        else:
+            # Matrix case:
+            raise NotImplementedError
```

Comment:
And this.

Reply:
Testing `NotImplementedError`?!?!?!?!?!?

Reply:
Yes. To make sure that the Exception is raised. Plenty of cases of that in the tests. Use `raises`.

The thing is that if an exception cannot be reached, as possibly the one above, the code shouldn't be there. It is much more convenient to add a test identifying a case where the Exception is thrown compared to a few years later trying to figure out if that branch of the code actually can be reached or if subsequent changes have effectively killed the code. Hence, it makes sense to add tests for all branches of the code.

Reply:
I really don't see the point of testing this case. The code will be added in the future.

Reply:
The purpose of tests is primarily to avoid future edits to introduce bugs or remove supported features. It does not make any sense to test unimplemented code blocks.

Reply:
It makes sense to test that the conditions that are expected to lead to that code block do lead there. The test that reaches that code also then serves as documentation for what is suposed to be there.